<?xml version="1.0"?>
<launch>
  <!-- configuration to be passed in -->
  <arg name="slamconfigFile" default="$(find aa_utils)/config/MH1-sptam-1803-slam.yaml" />
  <arg name="slamReaderconfigFile" default="$(find aa_utils)/config/MH1-sptam-1803-reader.yaml" />
  <arg name="gtReaderconfigFile" default="$(find aa_utils)/config/MH1-sptam-1803-gtreader.yaml" />
  <arg name="publisherconfigFile" default="$(find aa_utils)/config/MH1-sptam-1803-publisher.yaml" />

  <!-- Set use_sim_time true for datasets-->
  <param name="use_sim_time" value="true"/>
  <!-- static_transform_publisher x y z yaw pitch roll frame_id child_frame_id period_in_ms -->
  <node pkg="tf" type="static_transform_publisher" name="camera_broadcaster" args="0 0 0 0 0 0 base_link left_camera 30"/>
  <!-- Ground Truth path publisher. Transforms poses into path. -->

  <!-- Camera info publisher. Publishes CameraInfo for rectification -->
  <node pkg="aa_utils" type="testcamera_publisher.py" name="camera_info_pub">
    <rosparam command="load" file="$(arg publisherconfigFile)"/>
    <param name="pub_topic_r" value="/right/camera_info"/>
    <param name="sub_topic_r" value="/cam1/image_raw"/>
    <param name="pub_topic_l" value="/left/camera_info"/>
    <param name="sub_topic_l" value="/cam0/image_raw"/>
  </node>

  <!-- Stereo rectification node -->
  <node pkg="stereo_image_proc" type="stereo_image_proc" name="image_rect_node">
    <param name="queue_size" value="20"/>
    <param name="approximate_sync" value="true"/>
    <remap from="/left/image_raw" to="/cam0/image_raw"/>
    <remap from="/right/image_raw" to="/cam1/image_raw"/>
  </node>

  <!-- Main S-PTAM node -->
  <node pkg="sptam" type="sptam_node" name="sptam" output="screen">
    <!-- Read S-PTAM parameters file -->
    <rosparam command="load" file="$(arg slamconfigFile)"/>
    <param name="approximate_sync" value="true"/>
    <param name="camera_frame" value="/left_camera"/>
    <!-- Remap topics  -->
    <remap from="/stereo/left/image_rect" to="/left/image_rect"/>
    <remap from="/stereo/right/image_rect" to="/right/image_rect"/>
    <remap from="/stereo/left/camera_info" to="/left/camera_info"/>
    <remap from="/stereo/right/camera_info" to="/right/camera_info"/>
  </node>

  <!-- Pose reader. Reads pose/path/position/transform/marker and stores it into csv
  for error calculation. -->
  <node pkg="aa_utils" type="pyreader.py" name="path_to_csv_node">
    <rosparam command="load" file="$(arg slamReaderconfigFile)"/>
  </node>

  <!-- Ground truth reader. Reads ground truth and stores it into csv
  for error calculation. -->
  <node pkg="aa_utils" type="pyreader.py" name="gt_to_csv_node">
    <rosparam command="load" file="$(arg gtReaderconfigFile)"/>
  </node>

  <!-- Play the rosbag file to start -->
  <node pkg="rosbag" type="play" name="rosbag" args=" --clock $(arg bagFile)" />
</launch>
